core/drivers/firewall/stm32_risaf.c

```diff
@@ -555,10 +557,133 @@ stm32_risaf_pm(enum pm_op op, unsigned int pm_hint,
 	return res;
 }
 
+static TEE_Result stm32_risaf_acquire_access(struct firewall_query *fw,
+					     paddr_t paddr, size_t size,
+					     bool read, bool write)
+{
+	struct stm32_risaf_instance *risaf = NULL;
+	struct stm32_risaf_region *region = NULL;
+	uint32_t cidcfgr = 0;
+	unsigned int i = 0;
+	uint32_t cfgr = 0;
+	vaddr_t base = 0;
+	uint32_t id = 0;
+
+	assert(fw->ctrl->priv && (read || write));
+
+	if (paddr == TZDRAM_BASE && size == TZDRAM_SIZE)
+		return TEE_SUCCESS;
+
+	risaf = fw->ctrl->priv;
+	base = risaf_base(risaf);
+
+	if (fw->arg_count != 1)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/*
+	 * RISAF region configuration, we assume the query is as
+	 * follows:
+	 * firewall->args[0]: Region configuration
+	 */
+	id = _RISAF_GET_REGION_ID(fw->args[0]);
+
+	if (!id || id >= risaf->pdata.nregions)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < risaf->pdata.nregions; i++) {
+		if (id == _RISAF_GET_REGION_ID(risaf->pdata.regions[i].cfg)) {
+			region = &risaf->pdata.regions[i];
+			if (region->addr != paddr || region->len != size)
+				return TEE_ERROR_GENERIC;
+			break;
+		}
+	}
+	if (!region)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	cfgr = io_read32(base + _RISAF_REG_CFGR(id));
+	cidcfgr = io_read32(base + _RISAF_REG_CIDCFGR(id));
+
+	/*
+	 * Access is denied if the region is disabled and OP-TEE does not run as
+	 * TDCID, or the region is not secure, or if it is not accessible in
+	 * read and/or write mode, if requested, by OP-TEE CID.
+	 */
+	if ((!(cfgr & _RISAF_REG_CFGR_BREN) && !is_tdcid) ||
+	    !(cfgr & _RISAF_REG_CFGR_SEC) ||
+	    (cidcfgr &&
```

Comment:
If `cidcfgr`is 0, then access to CID1 is denied. I think line 614 should be removed and let the 2 lines below test access rights.

Reply:
Actually, I should add a check on  whether _RISAF_REG_CFGR_BREN is enabled. Else, the CID filtering values are not taken into account (case where  _RISAF_REG_CFGR_BREN is set but we're not TDCID).

Reply:
Refactored to fix some cases and make it clearer